### ISSUE TITLE
diffusion coefficient from rmsd or velocity autocorrelation function 

### ISSUE THREAD

**Issue by micaela-matta** [MDAnalysis/mdanalysis#2187]
I think it would be great to implement the calculation of diffusion coefficients using the Green-Kubo and/or Einstein formulas.

This feature is among [gromacs](http://www.gromacs.org/Documentation/How-tos/Diffusion_Constant) analysis tools and implemented in [cpptraj](https://github.com/Amber-MD/cpptraj/blob/master/src/Action_VelocityAutoCorr.cpp)/[pytraj](https://amber-md.github.io/pytraj/latest/_api/pytraj.all_actions.html#pytraj.all_actions.velocityautocorr) [ however, the latter is not well documented].

**Comment by mattwthompson** [comment]
I can take a stab at doing the Einstein half of this. I don't do much Green-Kubo and I'm not familiar with much more than the basics of it.

Somewhat related, ionic conductivity falls out of MSDs fairly easily either with Nernst-Einstein or Einstein-Helfand. Not sure this would be useful to tons of people but I have been using it lately.

**Comment by richardjgowers** [comment]
@mattwthompson this would be very welcome!  One thing I'm aware of is this: https://github.com/pdebuyl-lab/tidynamics

**Comment by kain88-de** [comment]
I worked a lot on diffusion in the last three years. Below are some  general comments to help implement *correct* calculation of diffusion coefficients.

---
tidydynamics and how to [calculate correlations](http://pdebuyl.be/blog/2016/correlators.html)

The blog post is interesting. It shows that FFT is a good method for calculation of the MSD. However, it means we have to store all coordinates of the trajectory in a numpy array. With the additional requirement of zero-padding for performance, it can be a memory hog. The blocking method is much more memory friendly but it requires to iterate over the trajectory multiple times. The later might not have a huge impact if we calculate the MSD of several molecules at a time.

---

[pydiffusion](https://github.com/bio-phys/pydiffusion) rotational and translational diffusion coefficients.  The rotational diffusion module is state of the art but requires preprocessing the trajectories to remove all translational diffusion artifacts. This is true for all rotational diffusion analytic codes. The translational diffusion calculations are similar to tidydynamics. However, I have included a method to determine the translational diffusion in an arbitrary protein fixed coordinate system. The later is of interest but I mainly used it to validate diffusion algorithms.

---

[A previous PR](https://github.com/MDAnalysis/mdanalysis/pull/1034)

We had a PR for the addition of MSD calculations a few years ago. The algorithms looked quite interesting. I never fully understood it but it looked like it can to a blocked correlation analysis with a single pass over the trajectory. It's worth having a look into it.

---
Diffusion coefficients in Periodic Boundary Systems

The Diffusion coefficient is heavily influenced by PBC that we often use in simulations. Any implementation of a diffusion coefficient calculation in MDAnalysis *should include PBC corrections*. Luckily the corrections only depend on the simulation box size. 

[membranes](https://github.com/bio-phys/memdiff)
[translation](https://pubs.acs.org/doi/abs/10.1021/jp0477147)
[rotation](https://pubs.acs.org/doi/abs/10.1021/acs.jpclett.8b01090)

For membranes in particular a huge warning sign should be added because for flat boxes the observed diffusion coefficient can grow indefinitely. 

---

Some general remarks about the calculation of diffusion coefficients.

Diffusion coefficients are calculated from correlation functions. The MSD is a correlation function. Unfortunately, correlation functions are notoriously difficult to calculate.  To reduce sampling noise it should be possible to average the MSD from several simulations or from different proteins of the same simulation.

**Comment by kain88-de** [comment]
If you have questions about some details you can ping me again on this issue.

**Comment by micaela-matta** [comment]
thanks a lot. I am only starting to dip my feet in diffusion-related problems.

**Comment by ranadevyesh** [comment]
> I can take a stab at doing the Einstein half of this. I don't do much Green-Kubo and I'm not familiar with much more than the basics of it.
> 
> Somewhat related, ionic conductivity falls out of MSDs fairly easily either with Nernst-Einstein or Einstein-Helfand. Not sure this would be useful to tons of people but I have been using it lately.

Yes conductivity would be very useful to me. How have you implemented this calculation?

**Comment by orbeckst** [comment]
The current development version (and upcoming 2.0.0) has [analysis.msd](https://docs.mdanalysis.org/2.0.0-dev0/documentation_pages/analysis/msd.html) (does not do Green-Kubo, though).

**Comment by kain88-de** [comment]
Note that there have recently been [papers](https://arxiv.org/abs/2003.09205) about the [systematic](https://aip.scitation.org/doi/full/10.1063/5.0008312) errors in self-diffusion coefficient calculation.  The authors also provide a [python package](https://github.com/bio-phys/DiffusionGLS) to remove the errors. The package is GPL3 licensed, but I'm sure if we ask nicely we can include it as GPL2 in MDAnalysis.

**Comment by xhgchen** [comment]
We now have a working implementation of Green-Kubo self-diffusivity at https://github.com/MDAnalysis/transport-analysis

I will update this comment when the package is officially released, but if anyone needs it before then, I would encourage cloning the repo or using the source code. The package uses [standard MDAnalysis units](https://userguide.mdanalysis.org/stable/units.html). The docs can be found [here](https://transport-analysis.readthedocs.io/en/latest/) and the reproduction blog post can be found [here](https://xhgchen.github.io/posts/reproduction-and-beyond-gsoc/). Please note that the analysis I performed in that blog post did not use standard MDAnalysis units because I was using OpenMM for the simulation and had to enter the velocities manually as a result.

Thanks to everyone at MDAnalysis for working with me and making this possible!
